dartagnan/src/main/java/com/dat3m/dartagnan/utils/visualization/ExecutionGraphVisualizer.java

```diff
@@ -99,17 +107,46 @@ private ExecutionGraphVisualizer addReadFrom(ExecutionModel model) {
         return this;
     }
 
+    private ExecutionGraphVisualizer addFromRead(ExecutionModel model) {
+
+        graphviz.beginSubgraph("FromRead");
+        graphviz.setEdgeAttributes("color=orange");
+        for (Map.Entry<EventData, EventData> rw : model.getReadWriteMap().entrySet()) {
+            EventData r = rw.getKey();
+            EventData w = rw.getValue();
+
+            if (ignore(r) || ignore(w)) {
+                continue;
+            }
+
+            for (List<EventData> co : model.getCoherenceMap().values()) {
+                for (int i = 2; i < co.size(); i++) {
+                    // We skip the init writes
+                    EventData w1 = co.get(i - 1);
+                    EventData w2 = co.get(i);                    
+                    if (ignore(w1) || ignore(w2) || !w.equals(w1) || frFilter.test(r, w2)) {
+                        continue;
+                    }
+                    
+                    appendEdge(r, w2, model, "label=fr");
+                }
+            }
+        }
```

Comment:
This is unnecessarily complicated. `model.getCoherenceMap` returns a map which gives you the coherence per address.
You can just do `model.getCoherenceMap().get(w.getAccessedAddress())` to get the total order that contains "w".
This list is ordered, so you just need to find the index of `w` in this list and then go over all successors.

```
List<EventData> co = model.getCoherenceMap().get(w.getAccessedAddress());
List<EventData> coLaterWrites = co.subList(co.indexOf(w) + 1, co.size());
for (EventData w2 : coLaterWrites) {
     if (!ignore(w2) && !frFilter.test(r, e2)) {
          appendEdge(r, w2, model, "label=fr");
     }
}
```

That being said, it might be good to leave out the transitive edges and just put the edge to the immediate co-successor (this avoids a quadratic number of edges).

Reply:
The graph is unreadible with the transitive ones, so I will only keep the edge to the immediate co-successor